### PR TITLE
Finished task, try-1.

### PR DESCRIPTION
**Pull request by slava-moiseev-kramatorsk**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private final int capacity = 10;
+    private final int changeStep = 1;
+    private final int zeroIndex = 0;
+    private T[] userArray;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:9]:
```suggestion
    private T[] values;
```
```suggestion
    private T[] elements;
```
```suggestion
    private T[] data;
```
don't include data type in the variable's name

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private final int capacity = 10;
+    private final int changeStep = 1;
+    private final int zeroIndex = 0;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:8]: remove these fields. 
we can obtain capacity from array

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,114 @@
...
+        }
+        userArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < zeroIndex || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index of element");
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:61]: repeated code,extract to a private method, include index and size in the message. You can hardcode 0 value for zeroIndex

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,114 @@
...
     public T remove(T element) {
-        return null;
+        T removedElement;
+        int indexRemove = findByValue(element);
+        if (indexRemove != -1) {
+            removedElement = userArray[indexRemove];
+            remove(indexRemove);
+        } else {
+            throw new NoSuchElementException("Element not found");
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:78]: invert condition, it's better to throw exception early, and unwrap the rest of the code

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,114 @@
...
+            }
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(userArray, zeroIndex, newElements, zeroIndex, size);
+            userArray = (T[]) newElements;
+        }
+    }
+
+    private int findByValue(T element) {
+        for (int i = 0; i < size; i++) {
+            if (element == null && userArray[i] == null
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:107]:
```suggestion
            if (element == userArray[i]
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,114 @@
...
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < zeroIndex || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index of element");
+        } else {
+            return userArray[index];
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:46]: we never need else after throwing, `throws`  exits the method immediately

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,114 @@
...
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement;
+        int indexRemove = findByValue(element);
+        if (indexRemove != -1) {
+            removedElement = userArray[indexRemove];
+            remove(indexRemove);
+        } else {
+            throw new NoSuchElementException("Element not found");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:77]: include element in the message

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,114 @@
...
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void changeCapacity(int actualCapacity) {
+        int oldCapacity = userArray.length;
+        if (actualCapacity > oldCapacity) {
+            int newCapacity = oldCapacity >> 1;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:95]: Better replace with simple multiplication for better readability and create a GROW_FACTOR constant

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,114 @@
...
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void changeCapacity(int actualCapacity) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:92]: remove argument, you always call it with the same value

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job! Read the comments to find out how to improve your code 👇

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,114 @@
...
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/ArrayList.java:37]: You can use for each loop here

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,114 @@
...
+    }
+
+    private void changeCapacity() {
+        int actualCapacity = size + 1;
+        int oldCapacity = data.length;
+        if (actualCapacity > oldCapacity) {
+            int newCapacity = (int) (oldCapacity * GROW_FACTOR);
+            if (newCapacity < actualCapacity) {
+                newCapacity = actualCapacity;
+            }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/ArrayList.java:91]: Grow your array by grow_factor, don`t add every time +1 just use default capacity and clean code after this

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 10 above)
Comment (slava-moiseev-kramatorsk) [reply src/main/java/core/basesyntax/ArrayList.java:37]: image
Ні, не можу.

File: src/main/java/core/basesyntax/ArrayList.java (hunk 11 above)
Comment (slava-moiseev-kramatorsk) [reply src/main/java/core/basesyntax/ArrayList.java:91]: Не розумію як обійтись без "+1"  це перевірка на розмір перед додаванням нового єлементу. Замінив на DEFAULT_CAPACITY. Чи малось на увазі  "1" зробити константою щоб було - 
private static final double GROW_FACTOR = 1;
.
.
"int actualCapacity = size + GROW_FACTOR;

File: src/main/java/core/basesyntax/ArrayList.java (hunk 10 above)
Comment (okuzan) [reply src/main/java/core/basesyntax/ArrayList.java:38]: та, там кастомний ліст, і в нього нема foreach , все ок

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,48 +1,110 @@
...
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void changeCapacity() {
+        if (size == data.length) {
+            int newCapacity = (int) (data.length * GROW_FACTOR);
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(data,0, newElements,0, size);
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:89]:
```suggestion
            System.arraycopy(data, 0, newElements, 0, size);
```
don't forget to autoformat your files before committing with Ctrl Alt L

**Review by okuzan — APPROVED** [review]
Good job! keep it up 💪

**Pull request closed without merging**